### DISCUSSION TITLE
With-msw example, If api call was made inside a useEffect block in first render, msw would be initialized after the api call, resulting in 404 response.

### DISCUSSION THREAD

**Discussion by VinceCYLiao** [vercel/next.js#39695] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 19.6.0: Thu Oct 29 22:56:45 PDT 2020; root:xnu-6153.141.2.2~1/RELEASE_X86_64
    Binaries:
      Node: 14.16.1
      npm: 7.24.2
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.2.4
      eslint-config-next: N/A
      react: 17.0.2
      react-dom: 17.0.2

### Which example does this report relate to?

with-msw

### Describe the Bug

In pages/index.tsx, if I move the fetch api call into a useEffect block to make an api call when the page first rendered, a 404 Not Found response will be returned.

After Inspecting console, I found out it was because the api call was made before the msw initialized.
<img width="490" alt="image" src="https://user-images.githubusercontent.com/46563912/184270986-a69c0456-8864-4cf1-b3f8-a439151f890e.png">

If I revert the mocks/index.ts to the old way using require to load server and browser module, msw will be properly initialized before the api call inside useEffect.

### Expected Behavior

Should get a status code 200 response with mock data.

### To Reproduce

Move the fetch api call into a useEffect block like below:
<img width="623" alt="image" src="https://user-images.githubusercontent.com/46563912/184270084-bff3bdec-5677-4d49-b078-927d7153e31f.png">

**Comment by okinawaa** [comment]
me too..

**Comment by souhailmerroun** [comment]
same, no updates?

**Comment by kira-0521** [comment]
me too...

It can be reproduced by using setTimeOut and waiting 1 second.

```typescript
useEffect(() => {
    setTimeout(() => {
      fetchTodos().then((todos) => {
        setTodoList(todos)
      })
    }, 1000)
  }, [])
```

**Comment by kevin-dev71** [comment]
i have the same problem.... any solutions?

**Comment by wesley-korput-maersk** [comment]
I have the same issue.

**Comment by LucasRMP** [comment]
I'm having the same problem

**Comment by souhailmerroun** [comment]
I ended up not using msw.

for those interested, I settled with https://github.com/ctimmerm/axios-mock-adapter, it works well when using axios for all the api calls server & client in my app.

**Comment by NikoKaranatsios** [comment]
Hello everyone, 

I was having the same issue and after some back and forth I was able to solve it. 
I use msw in a NextJS project. I had an additional bug, where msw was telling me, that it is not able to run in a non-browser environment. 

Both where solved, after I **awaited the initialization of my initMocks(); function** as well as the the dynamic import itself. 

in my _app.js file outside of the MyApp Component: 
![image](https://user-images.githubusercontent.com/43808737/211827989-ad893a32-5632-4144-bae3-991172cdd3c4.png)

in the /mocks/index.js file:
![image](https://user-images.githubusercontent.com/43808737/211828344-c3b0b8e7-63bf-4984-b8fb-722267e3f304.png)

this only works if you allow "topLevelAwait" in your webpack config like this: 
![image](https://user-images.githubusercontent.com/43808737/211828820-7b3ca877-8772-4e8c-a8ba-d7ab973d2a05.png)


By doing all this your code will wait for the mock service worker to be initialized before it makes any other requests.

**Reply by souhailmerroun** [reply]
Does it takes a lot to load, or barely noticable?

**Reply by NikoKaranatsios** [reply]
I was not able to tell a difference :)

**Reply by wesley-korput-maersk** [reply]
In the end I went for this solution: https://github.com/mswjs/msw/discussions/1049#discussioncomment-1941348

**Reply by souhailmerroun** [reply]
thanks ! i will try this later 🙏
